Argparse

ProvVlad opened the pull request:
Add CLI parser (argparse module) and add feature "read URL from file".

ProvVlad committed:
Add CLI parser, prepare to feature 'read URL from file' so change check_cli_arg function and add terminate script

ProvVlad committed:
Add feature "read url from file"

Shuk0 commented on scrapper.py:
```diff
@@ -1,35 +1,81 @@
-        2:"Argument have to begin from 'https://krisha.kz'.",
-        3:"Too many arguments. It's expected just one argument which is URL to be grabbed.",
-        4:"Page is not available. Check connection and restart the script."}
+def cli_parser() -> argparse.ArgumentParser:
```
@ProvVlad , let's rename it using verb

Shuk0 commented on scrapper.py:
```diff
@@ -1,35 +1,81 @@
+    """
+
+    cli_parser = argparse.ArgumentParser(description='Scrapper from krisha.kz')
+    cli_parser.add_argument('url', default=None, nargs='?',
```
@ProvVlad , please get acquainted with this argparse capability https://docs.python.org/3/library/argparse.html#mutual-exclusion. I suggest to change parser initialization in the following way:
`cli_parser = argparse.ArgumentParser(description='Scrapper from krisha.kz')`

`group = cli_parser.add_mutually_exclusive_group(required=True)`
`group.add_argument('url', default=None, nargs='?', help='site url for scrapping. Have to start with symbols "https://krisha.kz"')`

`group.add_argument('-ff', '--fromfile',  dest='file_with_url', help='''file with site url for scrapping.                            You have to specify the full path to file or file have to be on the same folder''')`

`args = cli_parser.parse_args()`

Shuk0 commented on scrapper.py:
```diff
@@ -1,35 +1,81 @@
     
-    print(OUTPUT_MESSAGES[exit_code])
-    exit(1)
+    return [args.url, args.file_with_url]
```
@ProvVlad , let's change return value to args instead of returning list of arguments.

Shuk0 commented on scrapper.py:
```diff
@@ -1,35 +1,81 @@
-    exit(1)
+    return [args.url, args.file_with_url]
 
+def check_cli_arg(command_line_arguments: list) -> list:
```
@ProvVlad , taking into account these capabilities https://docs.python.org/3/library/argparse.html#mutual-exclusion I suggest to change this function:
`def check_cli_arg(command_line_arguments) -> str:
    """ Function get command line arguments and check:
            - that URL or PATH to file with URL was entered in a command line;
            - that URL and PATH to file with URL didn't entered in a command line together.
    """

    if command_line_arguments.file_with_url:
        url = extract_url_from_file(command_line_arguments.file_with_url)
    else:
        url = command_line_arguments.url
    return check_url(url)`

Shuk0 commented on scrapper.py:
```diff
@@ -1,35 +1,81 @@
     """
+
+    OUTPUT_MESSAGES = {
+        1 : '''You missed an argument. It's exspected an URL or path to file with URL that should be grabbed. 
```
@ProvVlad , please revise this function taking into account changes in check_cli_args function.

Shuk0 commented on scrapper.py:
```diff
@@ -38,8 +84,8 @@ def grab_html_page(accepted_url: str, page_number: int) -> BeautifulSoup:
         page = requests.get(url, timeout=0.5)
-    except requests.ConnectionError as e:
-        terminate_script(4)
+    except requests.ConnectionError:
```
@ProvVlad , please, don't forget to handle ReadTimeout exception.

Shuk0 commented on scrapper.py:
```diff
@@ -96,7 +142,7 @@ def get_number_of_options(soup: BeautifulSoup) -> int:
         numbers_of_options = int(content_of_div_with_number.encode("ascii", "ignore")) # number of options
     except TypeError:
-        print("Can't parse number of options as extracted text is the following" + numbers_of_options)
+        print("Can't parse number of options as extracted text is the following" + content_of_div_with_number)
```
@ProvVlad , let's replace print() with terminate_script() to make error handling unified.

Shuk0 commented on scrapper.py:
```diff
@@ -96,7 +142,7 @@ def get_number_of_options(soup: BeautifulSoup) -> int:
+        print("Can't parse number of options as extracted text is the following" + content_of_div_with_number)
 
     return numbers_of_options
 
```
@ProvVlad , please rename function get_cost_per_sq_m to calculate_cost_per_sq_m. I'm asking you to do this, because getters and setters are specific naming with a corresponding semantics (https://www.geeksforgeeks.org/getter-and-setter-in-python/). Also, verb get means that you extract get from some source, for instance, from a file. But in this particular function verb calculate is more applicable.

Shuk0 commented on scrapper.py:
```diff
@@ -56,11 +102,11 @@ def get_data(soup: BeautifulSoup, data: dict) -> int:
-    for each in text_data:
+    for each in list_with_data:
         # parse price of flat
         text_with_one_flat_price = ''.join(each.find('div',
```
@ProvVlad m why do you use "".join() function? is it a "pythonic" style of casting to String type? It would be great if you share a link to some doc that explains this.

ProvVlad committed:
Renamed "cli_parser" function to "parse_cli_arg" and add to function mutual group.
Revise check_cli_arg function.
Revise terminate_script function
Renamed "get__cost_per_sq_m" to "calculate_cost_per_sq_m".

ProvVlad committed:
Import 2 new modules: random (randint) and time (sleep)
grab_html_page: ADD timeout and new try if have a connection or readtimeout error
get_data: FIX change 'price' parsing. find all 'words' begining numbers and join them.
	  ADD offer's date parsing
main: FIX add random timeout (4-8 sec) into loop.
